File: Tests/SwitchThemesNXTests/RealTests.cpp
#include "CppUnitTest.h"
#include "../../SwitchThemesNX/source/SwitchThemesCommon/SwitchThemesCommon.hpp"
#include "../../SwitchThemesNX/source/SwitchThemesCommon/Layouts/Bflan.hpp"
#include "../../SwitchThemesNX/source/SwitchThemesCommon/Layouts/Bflyt/Bflyt.hpp"
#include "picosha2.h"
#include "Util.hpp"
#include "../../SwitchThemesNX/source/SwitchThemesCommon/SarcLib/Sarc.hpp"

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace SwitchThemesNXTests
{
	TEST_CLASS(RealTests)
	{
	public:
		#define MAKE_TEST_SZS(x) TEST_METHOD(x) { ProcessSzs(#x); }
		
		MAKE_TEST_SZS(ResidentMenu)
		
		MAKE_TEST_SZS(Entrance)
		
		MAKE_TEST_SZS(Notification)
		
		MAKE_TEST_SZS(Flaunch)
		
		MAKE_TEST_SZS(Set)

	private:
		void CompareSarc(const SARC::SarcData& a, const SARC::SarcData& b)
		{
			if (a.files.size() != b.files.size())
				throw std::runtime_error("");

			for (const auto &[key, value] : a.files)
			{
				if (!b.files.count(key))
					throw std::runtime_error("file missing: " + key);

				if (b.files.at(key) != value)
					throw std::runtime_error("file different: " + key);
			}
		}
		
		void ProcessSzs(const std::string& name)
		{
			auto src = SARC::Unpack(Yaz0::Decompress(Util::ReadTestData("Source/" + name + ".szs")));
			auto exp = SARC::Unpack(Yaz0::Decompress(Util::ReadTestData("Expected/" + name + ".szs")));
		
			std::string lyt = Util::ExistsTest("Source/" + name + ".json") ? 
				Util::ReadTestString("Source/" + name + ".json") : "";

			SwitchThemesCommon::SzsPatcher p(std::move(src));
			if (!p.PatchMainBG(DDS))
				throw std::runtime_error("");

			if (lyt != "")
			{
				auto l = Patches::LoadLayout(lyt);
				if (!p.PatchLayouts(l))
					throw std::runtime_error("");
			}
			
			auto fin = p.GetFinalSarc();
			CompareSarc(exp, fin);
		}

		std::vector<u8> DDS = Util::ReadTestData("bg.dds");
	};
}

File: Tests/SwitchThemesCommonTests/RealTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using SARCExt;
using SwitchThemes.Common;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;

namespace SwitchThemesCommonTests
{
	[TestClass]
	public class RealTests
	{
		byte[] DDS = Util.ReadData("bg.dds");

		private void CompareSarc(SARCExt.SarcData a, SARCExt.SarcData b) 
		{
			Assert.AreEqual(a.Files.Count, b.Files.Count);

			foreach (var f in a.Files)
			{
				if (!b.Files.ContainsKey(f.Key))
					throw new Exception($"{f.Key} is missing in B");

				if (!b.Files[f.Key].SequenceEqual(f.Value))
					throw new Exception($"file {f.Key} is different in B");
			}
		}

		private void ProcessSzs(string name)
		{
			SarcData src = SARC.Unpack(ManagedYaz0.Decompress(Util.ReadData($"Source/{name}.szs")));			
			SarcData exp = SARC.Unpack(ManagedYaz0.Decompress(Util.ReadData($"Expected/{name}.szs")));

			string lyt = Util.Exists($"Source/{name}.json") ? Util.ReadString($"Source/{name}.json") : null;

			SzsPatcher patcher = new SzsPatcher(src);
			Assert.IsTrue(patcher.PatchMainBG(DDS));

			if (lyt != null)
			{
				var l = LayoutPatch.Load(lyt);
				patcher.PatchLayouts(l);
			}

			var final = patcher.GetFinalSarc();
			CompareSarc(final, exp);
		}

		[TestMethod]
		public void ResidentMenu() =>
			ProcessSzs("ResidentMenu");

		[TestMethod]
		public void Entrance() =>
			ProcessSzs("Entrance");

		[TestMethod]
		public void Notification() =>
			ProcessSzs("Notification");

		[TestMethod]
		public void Flaunch() =>
			ProcessSzs("Flaunch");

		[TestMethod]
		public void Set() =>
			ProcessSzs("Set");


		private void MaterialTexturesAssumption(string name)
		{
			SarcData src = SARC.Unpack(ManagedYaz0.Decompress(System.IO.File.ReadAllBytes(name)));

			foreach (var val in src.Files.Where(x => x.Key.EndsWith(".bflyt")))
			{
				var (n, f) = val;

				SwitchThemes.Common.Bflyt.BflytFile ff = new SwitchThemes.Common.Bflyt.BflytFile(f);

				if (ff.Mat1 != null)
				{
					foreach (var m in ff.Mat1.Materials)
						Assert.AreEqual(m.Textures.Length, m.TextureTransformations.Length);
				}
			}
		}

		[TestMethod]
		public void CheckMaterialTexturesAssumption() 
		{
			foreach (var f in System.IO.Directory.GetFiles(Util.GetPath("Source/")).Where(x => x.EndsWith(".szs")))
				MaterialTexturesAssumption(f);
		}
	}
}
